File: backend/services/progress_service.py
"""
Progress service — computes user progress analytics.
"""

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))

from core.database import (
    get_user_sessions,
    get_sessions_summary,
)


def get_progress(user_id: int) -> dict:
    """Compute full progress data for a user."""
    sessions = get_user_sessions(user_id)
    summary = get_sessions_summary(user_id)

    total_sessions = len(sessions)
    total_reps = sum(s.get("reps", 0) for s in sessions)
    avg_form = (
        sum(s.get("form_accuracy", 0) for s in sessions) / total_sessions
        if total_sessions > 0
        else 0
    )

    return {
        "total_sessions": total_sessions,
        "total_reps": total_reps,
        "avg_form_accuracy": round(avg_form, 1),
        "exercises_done": len(summary),
        "summary": summary,
        "recent_sessions": sessions[:10],
    }


def get_streak(user_id: int) -> int:
    """Compute the current exercise streak (consecutive days with sessions)."""
    from datetime import datetime, timedelta

    sessions = get_user_sessions(user_id)
    if not sessions:
        return 0

    # Get unique dates (just the date part)
    dates = set()
    for s in sessions:
        d = s.get("date", "")
        if d:
            dates.add(d[:10])

    # Sort dates descending
    sorted_dates = sorted(dates, reverse=True)
    today = datetime.now().strftime("%Y-%m-%d")

    streak = 0
    current = datetime.strptime(today, "%Y-%m-%d")

    for d_str in sorted_dates:
        d = datetime.strptime(d_str, "%Y-%m-%d")
        if d.date() == current.date() or d.date() == (current - timedelta(days=1)).date():
            streak += 1
            current = d
        else:
            break

    return streak

File: frontend/src/services/progressService.ts
import { apiRequest, isMockMode } from "./api";
import type { ProgressData, ProgressSummary } from "@/types/session";

export async function fetchProgress(): Promise<ProgressData> {
  if (isMockMode()) {
    return {
      total_sessions: 24,
      total_reps: 240,
      avg_form_accuracy: 82.5,
      exercises_done: 6,
      summary: [
        { exercise: "squat", total_sessions: 8, total_reps: 96, avg_form: 85 },
        { exercise: "lunges", total_sessions: 5, total_reps: 50, avg_form: 88 },
        { exercise: "shoulder_raises", total_sessions: 4, total_reps: 60, avg_form: 78 },
        { exercise: "calf_raises", total_sessions: 3, total_reps: 45, avg_form: 90 },
        { exercise: "knee_raises", total_sessions: 2, total_reps: 20, avg_form: 80 },
        { exercise: "tree_pose", total_sessions: 2, total_reps: 10, avg_form: 82 },
      ],
      recent_sessions: [
        { id: 1, user_id: 1, exercise: "squat", reps: 12, form_accuracy: 85, duration_sec: 180, notes: "", date: "2025-01-15 10:30:00" },
        { id: 2, user_id: 1, exercise: "lunges", reps: 10, form_accuracy: 90, duration_sec: 150, notes: "", date: "2025-01-14 09:15:00" },
        { id: 3, user_id: 1, exercise: "shoulder_raises", reps: 15, form_accuracy: 75, duration_sec: 120, notes: "", date: "2025-01-13 11:00:00" },
      ],
    };
  }
  return apiRequest<ProgressData>("/api/progress");
}

export async function fetchProgressSummary(): Promise<ProgressSummary> {
  if (isMockMode()) {
    return {
      total_sessions: 24,
      total_reps: 240,
      avg_form_accuracy: 82.5,
      exercises_done: 6,
      current_streak: 3,
    };
  }
  return apiRequest<ProgressSummary>("/api/progress/summary");
}
